### ISSUE TITLE
Regular crashes with "TLS record reception failed [gnutls error -54: Error in the pull function.]"

### ISSUE THREAD

**Issue by kamermans** [rsyslog/rsyslog#3915]
### Expected behavior
Rsyslog accepts TLS-encrypted RELP input and *does not* crash when a non-permanent TLS error is encountered.

### Actual behavior
Rsyslog accepts TLS-encrypted RELP input and *does* crash when a non-permanent TLS error is encountered.

### Steps to reproduce the behavior
I am able to reproduce this fairly easily in my environment, on more than one server (with different configs) and in more than one rsyslogd version (`v8.1904.0` and `v8.1910.0`).

### Environment
- rsyslog version: `v8.1910.0`
- platform: `Ubuntu 16.04`
- approx message rate: `6000/s`
- approx message size: `1000 bytes`

My architecture is as follows:
We have dozens of AWS instances that all send log messages via RELP+TLS to our rsyslog relay in AWS.  This relay simply forwards RELP+TLS messages to another rsyslog instance (aka our "rsyslog server") in our datacenter via RELP+TLS (this allows for buffering in case of an unstable connection to AWS).  We have been using this system for several years and it's been working great, but we switched from insecure RELP to RELP+TLS about 5 months ago, and have seen rsyslog crash dozens of times since then.  Lately it's been crashing 2+ times per week in both the relay and the server.

I know it is crashing because there is a process supervisor watching it and restarting it if it dies.  The process is then restarted and I can see "Starting rsyslogd" (from my init script) in the logs.

Before every crash, `rsyslogd` emits a message via `STDERR` like this:
```
imrelp[25140]: error 'TLS record reception failed [gnutls error -54: Error in the pull function.]', object  'lstn 25140: conn to clt 15.188.80.89/ec2-15-188-80-89.eu-west-3.compute.amazonaws.com' - input may not work as intended [v8.1910.0 try https://www.rsyslog.com/e/2353 ]
```

For example, here's a crash from a few hours ago, as seen from STDERR on the client (omrelp) and the relay (imrelp):
```
Client:
omrelp[rsyslog-relay:25140]: error 'TLS record write failed [gnutls error -53: Error in the push function.]', object  'conn to srvr rsyslog-relay:25140' - action may not work as intended [v8.1904.0 try https://www.rsyslog.com/e/2353 ]
omrelp[rsyslog-relay:25140]: error 'io error in RawSendCommand, session broken', object  'conn to srvr rsyslog-relay:25140' - action may not work as intended [v8.1904.0 try https://www.rsyslog.com/e/2353 ]
librelp error 10007[connection broken] forwarding to server rsyslog-relay:25140 - suspending  [v8.1904.0 try https://www.rsyslog.com/e/2291 ]

Relay:
imrelp[25140]: error 'TLS record reception failed [gnutls error -54: Error in the pull function.]', object  'lstn 25140: conn to clt 3.4.5.6/ec2-3-4-5-6.eu-west-3.compute.amazonaws.com' - input may not work as intended [v8.1910.0 try https://www.rsyslog.com/e/2353 ]
```

> Note that only the receiver crashes, the client eventually reconnects and continues sending messages.

Here is the sanitized relay (receiver) config:
`/etc/rsyslog.conf`
```
# Stats logging must be first!
module( load="impstats"
    interval="30"
    severity="7"
    log.syslog="off"
    log.file="/tmp/rsyslog-stats.log"
)
# End stats logging

# Can't figure out how to set these in RainerScript format
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
$RepeatedMsgReduction off
$FileOwner syslog
$FileGroup syslog
$FileCreateMode 0644
$DirCreateMode 0755
$Umask 0022
$PrivDropToUser syslog
$PrivDropToGroup syslog

global (
    maxMessageSize="64k"
    workDirectory="/var/spool/rsyslog"
    shutdown.enable.ctlc="on"
    net.aclResolveHostname="off"
)

# Super-verbose, be careful!
# $DebugFile /var/log/rsyslog-debug.log
# $DebugLevel 2

#module( load="imklog"
#    permitNonKernelFacility="on"
#)

module( load="imuxsock")
module( load="imudp"
    threads="2"
    timeRequery="8"
    batchSize="256"
)
module( load="imrelp" )
module( load="omrelp" )

input(  type="imuxsock"
    socket="/dev/log"
)

###############################
### Load other config files ###
###############################

$IncludeConfig /etc/rsyslog.d/*.conf
```

`/etc/rsyslog.d/10-forwarder.conf`
```
##################
### Main Queue ###
##################
main_queue(
    # [dlang]: some smallish value just so the worker isn't in a tight loop locking the main queue.
    # I'd try 500us as a starting point. that will limit your throughput to < 2000batches/sec *
    # 128 messages/batch or probably somewhere around 200K messages/sec
    queue.dequeueslowdown="500"
)

###########################
### Ruleset Partition 1 ###
###########################
ruleset(name="forwardPartition1"
    queue.type="LinkedList"

    # The number of messages that can be stored in RAM
    # Theoretical MAX: (15GB RAM / 1.5k msg) == 10,000,000
    # System has crashed with 1 queue at 7.8M
    # Keep this UNDER 6M
    queue.size="1000000"
    queue.highwatermark="900000"
    queue.lowwatermark="20000"

    queue.dequeueBatchSize="750"
    queue.saveOnShutdown="on"
    queue.workerThreads="1"


    # Setting these enables Disk-Assisted mode
    queue.fileName="something_queue"
    queue.spoolDirectory="/var/spool/rsyslog/something_part1"
    # The total amount of disk space that can be used to store
    #  messages that overflow the memory-based LinkedList queue
    # NOTE: queue.size has no effect on the disk-assisted storage
    queue.maxDiskSpace="35g"
    queue.maxFileSize="1g"
) {
    action( type="omrelp"
            name="fp1_omrelp"
            target="rsyslog-server"
            port="25140"
            tls="on"
            tls.caCert="/certs/ca.pem"
            tls.myCert="/certs/cert.pem"
            tls.myPrivKey="/certs/key.pem"
            tls.authmode="name"
            tls.permittedpeer=["rsyslog-server.mydomain.com", "rsyslog-relay.mydomain.com"]
            action.resumeRetryCount="-1"
            action.reportSuspension="on"
            action.reportSuspensionContinuation="on"
    )
}

######################
### Logging Inputs ###
######################
# TLS Encrypted TCP/RELP
input(  type="imrelp"
    port="25140"
    tls="on"
    tls.caCert="/certs/ca.pem"
    tls.myCert="/certs/cert.pem"
    tls.myPrivKey="/certs/key.pem"
    tls.authmode="name"
    tls.permittedpeer=["rsyslog-relay.mydomain.com", "rsyslog-client.mydomain.com"]
    ruleset="forwardPartition1"
)
```

Any ideas on how I can try to debug this further?  It seems like a random TLS failure during one of the 15 billion messages sent per month is somewhat expected, but it should not crash the entire process.

**Comment by alorbach** [comment]
Hi @kamermans , its difficult to debug seg faults further informations like gdb stacktrace. 
Can you try to create and analyse core files according to this site: 
https://www.rsyslog.com/doc/v8-stable/troubleshooting/troubleshoot.html

When we see a stacktrace of the crash, its easier to narrow the cause of the problem.

Alternative, I would recommend to switch from GNUTLS to OpenSSL driver. OpenSSL is much more verbose and helpful when it comes to error messages.

**Comment by kamermans** [comment]
Thanks @alorbach, I will try to get a stacktrace for you.  I imagine switching to openssl could work around the issue.  I'm working with the packages from `ppa:adiscon/v8-stable`:

```
rsyslog
rsyslog-kafka
rsyslog-relp
rsyslog-elasticsearch
rsyslog-mmjsonparse
rsyslog-imptcp
rsyslog-gnutls
rsyslog-openssl
```

I'll try to get it working without `rsyslog-gnutls`.  I'm not opposed to compiling it from scratch if need be, but I prefer to track the official stable repo if possible.

**Comment by davidelang** [comment]
we all prefer to have everyone running from the official stable repo if possible 
:-)

however, when we are tracking down a problem, it's a huge help if you can 
compile from scratch as that lets us try different things and confirm that the 
fix works for you before it gets put into the next stable repo. So it would only 
be a temporary thing to run a nightly or custom version

David Lang

**Comment by kamermans** [comment]
No problem, I'm happy to run a custom version if we need to dig deeper.  I ran out of time last night, but I already had `rsyslog-openssl` installed, so I switched my configs from `gnutls` to `openssl` by using this syntax:

```
module( load="imrelp" tls.tlslib="openssl" )
module( load="omrelp" tls.tlslib="openssl" )
```

Now it's been about 12 hours and I don't see any crashes yet.  I would expect it to crash this afternoon so I'll watch it and report back.
